posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+  simp only [↓reduceIte]
+
+example (T : Type) (P : Prop) (a b : T) : ite (P ∨ ¬ P) a b = a := by
+  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
+  simp only [↓reduceIte]
+
+example : ite FermatLastTheorem 1 2 = 1 := by
+  --This fails because `simp` can't reduce `FermatLastTheorem` to `True`
+  simp only [↓reduceIte]
+```
+
+Internally, this simproc is a small metaprogram does the following whenever an expression of the form `ite P a b` is encountered:
+- Call `simp` on `P` to get a simplified expression `P'` and a proof `r` that `P = P'`.
+**TODO Paul** check that the arguments in the proof are alright.
+- If `P' = True` then return the simplified expression `a` and the proof `ite_cond_eq_true r` that `ite P a b = a`
+- If `P' = False` then return the simplified expression `b` and the proof `ite_cond_eq_false r` that `ite P a b = b`
+- Otherwise, let `simp` continue the search.
+
+Source code is
+```lean
+builtin_simproc ↓ [simp, seval] reduceIte (ite _ _ _) := fun e => do
+  let_expr f@ite α c i tb eb ← e | return .continue
+  let r ← simp c
+  if r.expr.isTrue then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_true f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := tb, proof? := pr }
+  if r.expr.isFalse then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_false f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := eb, proof? := pr }
+  return .continue
+```
+
+Intuitively, one can think of this as a "parametric" lemma that allows one to vary the right hand side depending on the value of the left hand side.
+In the case of `reduceIte`, this allows us to combine `ite_cond_eq_true` and `ite_cond_eq_false` into a single procedure that `simp` can call.
+Notice that on can use `ite_cond_eq_true` (resp `ite_cond_eq_false`) instead of `↓reduceIte`, at the cost of introducing more steps in the simplification procedure:
+
+```
+set_option trace.Meta.Tactic.simp true in
+example : ite (1≠1) (2^3 * 0 * 50 * 50) (3^2) = 3^2 := by
+  -- 6 steps: 5 rewrites and one goal discharge
+  simp only [ite_cond_eq_false, ne_eq, not_true_eq_false]
+
+set_option trace.Meta.Tactic.simp true in
+example : ite (1≠1) (2^3 * 0 * 50 * 50) (3^2) = 3^2 := by
+  -- 4 rewrites
+  simp
+```
+
+### Computation: The `reduceDvd` simproc
+
+The `reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.
+
+```lean
+example : 3 ∣ 9 := by
+  simp only [Nat.reduceDvd]
+
+example : ¬ 2 ∣ 49 := by
+  simp only [Nat.reduceDvd, not_false_iff]
+
+example : a ∣ a * b := by
+  simp only [Nat.reduceDvd] --fails
+```
+
+Here the metaprogram run by `Nat.reduceDvd` does the following whenever an expression of the form `a ∣ b` (where `a, b` are natural numbers) is encountered:
+- Check that `∣` is the usual natural number division.
+- Try to extract explicit (literal) values for `a` and `b`.
```

Comment:
At this point "literal" is pretty meaningless to the reader, and "extract explicit values" is ambiguous too - it could mean "compute a and b" as well: this line has a pretty high chance of being confusing.

Reply:
I've made this line talk about numerals instead. I think that's plenty fine.